src/ttruntime.js

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+//     You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+//     Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+//     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//     See the License for the specific language governing permissions and
+// limitations under the License.
+
+const puppeteer = require('puppeteer');
+const yargs = require('yargs');
+
+const argv = yargs
+    .option('endpoint', {
+        alias: 'e',
+        description: 'Endpoint of running application to scan. Default: http://localhost:8080',
+        type: 'string',
+    })
+    .option('verbose', {
+        alias: 'v',
+        description: 'Verbose logging',
+        type: 'boolean'
+    })
+    .option('no-headless', {
+        alias: 'hl',
+        description: 'Open browser in the headless mode. Default: true',
+        type: 'boolean'
+    })
+    .option('path', {
+        alias: 'p',
+        description: 'Path to the tested project\'s root. Default: this project\'s root',
+        type: 'string',
+    })
+    .help()
+    .alias('help', 'h')
+    .argv;
+
+const violations = new Map();
+
+function parseViolation(data) {
+    const report = JSON.parse(data)['csp-report'];
+    if (report) {
+        const location = `${report['source-file']}:${report['line-number']}:${report['column-number']}`;
+        if (!violations.has(location)) {
+            violations.set(location, []);
+        }
+        violations.get(location).push(report);
+    }
+}
+
+async function intercept(page) {
+    const client = await page.target().createCDPSession();
+
+    await client.send('Fetch.enable', {
+        patterns: [
+            {requestStage: 'Response', resourceType: 'Document'},
+            {requestStage: 'Response', resourceType: 'Script'},
+            {requestStage: 'Response', resourceType: 'CSPViolationReport'},
+        ]
+    });
+
+    client.on('Fetch.requestPaused', async event => {
+        const { requestId, resourceType, request, responseHeaders } = event;
+        if (argv.verbose) {
+            console.log(`Intercepted ${request.url} {interception id: ${requestId}}`);
+        }
+        if (resourceType === 'CSPViolationReport') {
```

Comment:
I was a bit confused by this piece of code because you are mixing parsing violation reports and adding the header. Would it not be better to intercept the requests here ?

Reply:
Rename the `parseViolation` to `saveViolation` to make it more clear that it's only about saving the info from the response.